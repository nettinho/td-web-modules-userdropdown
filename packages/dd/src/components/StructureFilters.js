import _ from "lodash/fp";
import React from "react";
import { bindActionCreators, compose } from "redux";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { Dropdown } from "semantic-ui-react";
import { injectIntl, FormattedMessage } from "react-intl";
import routes from "../routes";
import {
  fetchStructures,
  clearStructureQueryFilters,
  addStructureSelectedFilter,
  openStructureFilter
} from "../routines";

const sortOrder = _.flow(
  _.toLower,
  _.deburr
);

const sortFilters = ({ structureFilters, selectedFilters, formatMessage }) =>
  _.flow(
    _.mapValues(_.filter(_.negate(_.isEmpty))),
    _.mapValues(_.sortBy(sortOrder)),
    _.omit(selectedFilters),
    _.entries,
    _.filter(([, v]) => v.length > 1),
    _.sortBy(([filterName]) =>
      sortOrder(formatMessage({ id: `filters.${filterName}` }))
    ),
    _.keyby(([_k]) => _k),
    _.mapValues(([, v]) => v)
  )(structureFilters);

export class StructureFilters extends React.Component {
  static propTypes = {
    structureFilters: PropTypes.object,
    intl: PropTypes.object,
    clearStructureQueryFilters: PropTypes.func,
    addStructureSelectedFilter: PropTypes.func,
    structureSelectedFilters: PropTypes.object,
    openStructureFilter: PropTypes.func
  };

  render = () => {
    const {
      structureFilters,
      clearStructureQueryFilters,
      intl: { formatMessage },
      structureSelectedFilters: { selectedFilters },
      addStructureSelectedFilter,
      openStructureFilter
    } = this.props;

    let filterOptions = sortFilters({
      structureFilters,
      selectedFilters,
      formatMessage
    });

    return (
      <React.Fragment>
        <Dropdown
          text={formatMessage({
            id: "structures.filters",
            defaultMessage: "Filters"
          })}
          icon="filter"
          floating
          labeled
          button
          className="icon"
          scrolling
          upward={false}
          onOpen={() => openStructureFilter("")}
        >
          <Dropdown.Menu>
            <Dropdown.Item onClick={clearStructureQueryFilters}>
              <em>
                <FormattedMessage
                  id="structures.filters.reset"
                  defaultMessage="(reset filters)"
                />
              </em>
            </Dropdown.Item>
            {Object.keys(filterOptions).map(filterName => (
              <Dropdown.Item
                key={filterName}
                text={formatMessage({ id: `filters.${filterName}` })}
                onClick={() => {
                  addStructureSelectedFilter(filterName);
                }}
              />
            ))}
          </Dropdown.Menu>
        </Dropdown>
      </React.Fragment>
    );
  };
}

const mapStateToProps = ({
  structureQuery: { filters },
  structureFilters,
  structureSelectedFilters
}) => ({
  structureFilters,
  active: !_.isEmpty(filters),
  structureSelectedFilters
});

const mapDispatchToProps = dispatch => ({
  ...bindActionCreators(
    {
      fetchStructures,
      clearStructureQueryFilters,
      addStructureSelectedFilter,
      openStructureFilter
    },
    dispatch
  )
});

export default compose(
  injectIntl,
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)(StructureFilters);
