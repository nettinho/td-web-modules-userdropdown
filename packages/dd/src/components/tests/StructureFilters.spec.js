import React from "react";
import { shallowWithIntl } from "intl-stub";
import { StructureFilters } from "../StructureFilters";

describe("<StructureFilters />", () => {
  const active = true;
  const structureSelectedFilters = {
    selectedFilters: [],
    openFilter: undefined
  };

  it("matches the latest snapshot", () => {
    const props = { active, structureSelectedFilters };
    const wrapper = shallowWithIntl(<StructureFilters {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
