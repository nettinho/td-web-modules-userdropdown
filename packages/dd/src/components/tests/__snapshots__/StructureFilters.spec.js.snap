// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StructureFilters /> matches the latest snapshot 1`] = `
<React.Fragment>
  <Dropdown
    additionLabel="Add "
    additionPosition="top"
    button={true}
    className="icon"
    closeOnBlur={true}
    deburr={false}
    floating={true}
    icon="filter"
    labeled={true}
    minCharacters={1}
    noResultsMessage="No results found."
    onOpen={[Function]}
    openOnFocus={true}
    renderLabel={[Function]}
    scrolling={true}
    searchInput="text"
    selectOnBlur={true}
    selectOnNavigation={true}
    text="structures.filters"
    upward={false}
    wrapSelection={true}
  >
    <DropdownMenu>
      <DropdownItem>
        <em>
          <FormattedMessage
            defaultMessage="(reset filters)"
            id="structures.filters.reset"
            values={Object {}}
          />
        </em>
      </DropdownItem>
    </DropdownMenu>
  </Dropdown>
</React.Fragment>
`;
